Draft: Keyboard SPI support

thejpster opened the pull request:
Adds support for reading the PS/2 keyboard buffers over SPI.

jonathanpallant committed:
Working on adding keyboard bytes to the SPI interface

thejpster committed:
Adds a basic keyboard FIFO.

Detegr commented on neotron-bmc-pico/.cargo/config.toml:
```diff
@@ -1,5 +1,5 @@
 [target.'cfg(all(target_arch = "arm", target_os = "none"))']
-runner = "probe-run --chip STM32F030K6Tx"
+runner = "probe-run --probe 0483:3748 --chip STM32F030K6Tx"
```
Committed by accident?

thejpster replied:
Yeah :(

Detegr commented on neotron-bmc-commands/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+impl Command {
+	pub fn parse(byte: u8) -> Option<Command> {
+		match byte {
+			0x00 => Some(Command::ProtocolVersion),
+			0x01 => Some(Command::FirmwareVersion),
+			0x10 => Some(Command::InterruptStatus),
+			0x11 => Some(Command::InterruptControl),
+			0x20 => Some(Command::ButtonStatus),
+			0x21 => Some(Command::SystemTemperature),
+			0x22 => Some(Command::SystemVoltage33S),
+			0x23 => Some(Command::SystemVoltage33),
+			0x24 => Some(Command::SystemVoltage55),
+			0x25 => Some(Command::PowerControl),
+			0x30 => Some(Command::UartBuffer),
+			0x31 => Some(Command::UartFifoControl),
+			0x32 => Some(Command::UartControl),
+			0x33 => Some(Command::UartStatus),
+			0x34 => Some(Command::UartBaudRate),
+			0x40 => Some(Command::Ps2KbBuffer),
+			0x41 => Some(Command::Ps2KbControl),
+			0x42 => Some(Command::Ps2KbStatus),
+			0x50 => Some(Command::Ps2MouseBuffer),
+			0x51 => Some(Command::Ps2MouseControl),
+			0x52 => Some(Command::Ps2MouseStatus),
+			0x60 => Some(Command::I2cBuffer),
+			0x61 => Some(Command::I2cFifoControl),
+			0x62 => Some(Command::I2cControl),
+			0x63 => Some(Command::I2cStatus),
+			0x64 => Some(Command::I2cBaudRate),
```
This looks a bit error prone. If you don't mind an additional dependency, something like [num_enum](https://docs.rs/num_enum/0.5.7/num_enum/) could be used to avoid writing the conversions manually.

thejpster replied:
That's a good idea, thanks!

thejpster replied:
It grinds my gears that you can't `match u8` and use a `repr(u8)` as the pattern match.

thejpster replied:
I also wonder if I should just re-export the key codes from the `pc-keyboard` crate. Saves a big conversion table inside the BIOS.

Edit: Oh wait, that's in the other repo. Never mind.

thejpster committed:
Use num_enum instead of hand-rolling a parse function.

Also fixes the unit tests for neotron-bmc-protocol.

jonathanpallant committed:
SPI interface is more robust.

Runs mostly OK at 2 MHz. Sometimes get the SPI RX buffer stuck two-bits offset. It probably needs a reset.
